stacktrace.go

```diff
@@ -20,23 +20,75 @@
 
 package zap
 
-import "runtime"
+import (
+	"runtime"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap/internal/bufferpool"
+)
+
+var (
+	_stacktraceIgnorePrefixes = []string{
+		"go.uber.org/zap",
+		"runtime.",
+	}
+	_stacktraceUintptrPool = sync.Pool{
+		New: func() interface{} {
+			return make([]uintptr, 64)
+		},
+	}
+)
 
 // takeStacktrace attempts to use the provided byte slice to take a stacktrace.
 // If the provided slice isn't large enough, takeStacktrace will allocate
 // successively larger slices until it can capture the whole stack.
-func takeStacktrace(buf []byte, includeAllGoroutines bool) string {
-	if len(buf) == 0 {
-		// We may have been passed a nil byte slice.
-		buf = make([]byte, 1024)
+func takeStacktrace() string {
+	buffer := bufferpool.Get()
+	programCounters := _stacktraceUintptrPool.Get().([]uintptr)
+	defer bufferpool.Put(buffer)
+	defer _stacktraceUintptrPool.Put(programCounters)
+
+	for {
+		n := runtime.Callers(2, programCounters)
+		if n < len(programCounters) {
+			programCounters = programCounters[:n]
+			break
+		}
+		// Do not put programCounters back in pool, will put larger buffer in
+		// This is in case our size is always wrong, we optimize to pul
+		// correctly-sized buffers back in the pool
+		programCounters = make([]uintptr, len(programCounters)*2)
 	}
-	n := runtime.Stack(buf, includeAllGoroutines)
-	for n >= len(buf) {
-		// Buffer wasn't large enough, allocate a larger one. No need to copy
-		// previous buffer's contents.
-		size := 2 * n
-		buf = make([]byte, size)
-		n = runtime.Stack(buf, includeAllGoroutines)
+
+	i := 0
+	for _, programCounter := range programCounters {
+		f := runtime.FuncForPC(programCounter)
+		name := f.Name()
+		if shouldIgnoreStacktraceName(name) {
```

Comment:
We've discussed this a lot, I think I'd like to merge this as is and go from there :)